unified-codes/apps/data-service/src/app/IdentityProvider.ts

```diff
@@ -0,0 +1,54 @@
+import JWT, { JWTToken } from './JWT';
+
+export interface IIdentityCredentials {
+  username: string;
+  password: string;
+}
+
+export interface IKeyCloakConfig {
+  baseUrl: string;
+}
+
+export class IdentityProviderError extends Error {
+  constructor(message: string) {
+    super(message);
+  }
+}
+
+export abstract class IdentityProvider {
+  abstract async getIdentityToken(credentials: IIdentityCredentials): Promise<JWTToken>;
+  abstract async verifyIdentityToken(token: JWTToken);
+}
+
+export class KeyCloakIdentityProvider extends IdentityProvider {
+  config: IKeyCloakConfig;
+
+  constructor(config: IKeyCloakConfig) {
+    super();
+    this.config = config;
+  }
+
+  async getIdentityToken(credentials: IIdentityCredentials) {
+    // Add header/body/whatever we need to pass the credentials?
+    const response = await fetch(this.config.baseUrl);
+    const token: JWTToken = JWT.parseResponse(response);
+    return token;
+  }
+
+  async verifyIdentityToken(token: JWTToken) {
+    const response = await fetch(this.config.baseUrl);
+    const payload = await response.json();
+    const { public_key: publicKey } = payload ?? {};
+
+    if (!publicKey) throw new IdentityProviderError('public key not found');
+
+    const secret = `
```

Comment:
Ah, that sucks. Maybe would be nice to wrap this in a class then, with a `toString` which adds the "padding"?

Reply:
actually, this would do the same thing: 

```const secret = `-----BEGIN PUBLIC KEY-----\n${publicKey}\n-----END PUBLIC KEY-----`;```

 That's not too bad is it?